lib/phoenix/controller.ex

```diff
@@ -146,14 +151,66 @@ defmodule Phoenix.Controller do
   """
   @spec json(Plug.Conn.t, term) :: Plug.Conn.t
   def json(conn, data) do
-    encoder =
-      Application.get_env(:phoenix, :format_encoders)
-      |> Keyword.get(:json, Poison)
-
+    encoder = get_json_encoder()
+    
     send_resp(conn, conn.status || 200, "application/json", encoder.encode_to_iodata!(data))
   end
 
   @doc """
+  Sends a JSON response with JSONP callback support.
+
+  It uses the configured `:format_encoders` under the `:phoenix`
+  application for `:json` to pick up the encoder module.
+
+  If the query params contain a key for the callback name, the 
+  response body is wrapped in an invocation of this callback.
+
+  The default key for the callback name is `"callback"`. It can be
+  overridden with the `:callback` option and is expected to be a 
+  string.
+
+  Only alphanumeric characters and underscore is allowed in the
+  callback name. Otherwise an exception is raised.
+
+  ## Examples
+
+      iex> jsonp conn, %{id: 123}, callback: "cb"
+
+  """
+  @spec jsonp(Plug.Conn.t, term, Keyword.t) :: Plug.Conn.t
+  def jsonp(conn, data, opts \\ []) do
+    callback_param = Keyword.get(opts, :callback, "callback")
+    callback = Dict.get(conn.query_params, callback_param)
+
+    if callback && String.length(callback) > 0 do
+      encoder = get_json_encoder()
+
+      allowed_chars = [?0..?9, ?A..?Z, ?a..?z, [?_]]
+      |> Enum.map(&Enum.to_list/1)
+      |> List.flatten()
+
+      callback_valid = callback
+      |> String.to_char_list()
+      |> Enum.all?(&(&1 in allowed_chars))
+
+      callback_valid || raise ArgumentError, "the callback name contains invalid characters"
+
+      body = encoder.encode_to_iodata!(data) 
+      |> IO.iodata_to_binary()
+
+      body = body
+      |> String.replace(<< 0x2028 :: utf8 >>, "\\u2028")
+      |> String.replace(<< 0x2029 :: utf8 >>, "\\u2029")
+      
+      body = "/**/ typeof " <> callback <> " === 'function' && " <> callback <> "(" <> body <> ");"
+
+      send_resp(conn, conn.status || 200, "text/javascript", body)
+    else
+      json conn, data
```

Comment:
The advantage is that it let's you define an API that works both, same domain and cross domain. When it's not cross domain you just leave away the callback and get a normal json response. Otherwise you would have to treat those two cases separately.